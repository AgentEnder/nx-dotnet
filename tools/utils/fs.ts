import { Workspaces } from '@nrwl/tao/src/shared/workspace';

import { readFileSync, statSync, writeFileSync } from 'fs';
import { join } from 'path';
import { format } from 'prettier';

export function existsSync(path: string) {
  let results;
  try {
    results = statSync(path);
  } catch {}
  return !!results;
}

export function readJson(path: string) {
  return JSON.parse(readFileSync(path).toString());
}

export function writeJson(path: string, object: any) {
  const contents = format(JSON.stringify(object, null, 2), {
    parser: 'json',
  });
  return writeFileSync(path, contents);
}

export function readWorkspaceJson() {
  return new Workspaces(join(__dirname, '../../')).readWorkspaceConfiguration();
}
