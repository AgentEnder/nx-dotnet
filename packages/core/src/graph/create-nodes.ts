import { CreateNodes, TargetConfiguration, workspaceRoot } from '@nx/devkit';

import { readFileSync } from 'fs';
import { dirname, resolve } from 'path';

import { readConfig } from '@nx-dotnet/utils';

import {
  GetBuildExecutorConfiguration,
  GetLintExecutorConfiguration,
  GetServeExecutorConfig,
  GetTestExecutorConfig,
} from '../models';

export const projectFilePatterns = readConfig().inferProjects
  ? ['*.csproj', '*.fsproj', '*.vbproj']
  : [];

export const registerProjectTargets = (projectFile: string) => {
  const targets: Record<string, TargetConfiguration> = {};
  const { inferProjectTargets } = readConfig();
  if (inferProjectTargets ?? true) {
    const projectFileContents = readFileSync(
      resolve(workspaceRoot, projectFile),
      'utf8',
    );
    if (projectFileContents.includes('Microsoft.NET.Test.Sdk')) {
      targets['test'] = GetTestExecutorConfig();
    }
    targets['build'] = GetBuildExecutorConfiguration(dirname(projectFile));
    targets['lint'] = GetLintExecutorConfiguration();
    targets['serve'] = GetServeExecutorConfig();
  }
  return targets;
};

// Used in Nx 16.8+
export const createNodes: CreateNodes | undefined = readConfig().inferProjects
  ? [
      `{${projectFilePatterns.join(',')}}`,
      (file) => {
        const root = dirname(file);

        // eslint-disable-next-line no-useless-escape -- eslint's wrong
        const parts = root.split(/[\/\\]/g);
        const name = parts[parts.length - 1].toLowerCase();

        const targets = registerProjectTargets(file);

        return {
          projects: {
            [name]: {
              name,
              root,
              type: 'lib',
              targets,
              tags: ['nx-dotnet'],
            },
          },
        };
      },
    ]
  : undefined;
