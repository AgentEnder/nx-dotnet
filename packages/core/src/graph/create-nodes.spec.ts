import type { ResolvedConfig } from '@nx-dotnet/utils';
import type configUtils = require('@nx-dotnet/utils/src/lib/utility-functions/config');

import * as fs from 'fs';

let configValues: Partial<ResolvedConfig> = {
  inferProjects: true,
};

jest.mock(
  '@nx-dotnet/utils/src/lib/utility-functions/config',
  () =>
    ({
      readConfig: () => configValues,
    } as typeof configUtils),
);

import { registerProjectTargets } from './create-nodes';

describe('infer-project', () => {
  beforeEach(() => {
    configValues = {
      nugetPackages: {},
      inferredTargets: {
        build: 'build',
        lint: 'lint',
        serve: 'serve',
        test: 'test',
      },
      ignorePaths: [],
      tags: ['nx-dotnet'],
    };
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should obey inferProjectTargets: false', () => {
    configValues.inferredTargets = false;
    jest.spyOn(fs, 'readFileSync').mockReturnValueOnce('<project></project>');

    expect(registerProjectTargets('libs/api/my.csproj')).toEqual({});
  });

  it('should generate build, lint, serve targets for projects', () => {
    jest.spyOn(fs, 'readFileSync').mockReturnValueOnce('<project></project>');

    const targets = registerProjectTargets('libs/api/my.csproj');
    expect(targets.build).toMatchSnapshot();
    expect(targets.lint).toMatchSnapshot();
    expect(targets.serve).toMatchSnapshot();
    expect(targets.test).not.toBeDefined();
  });

  it('should generate test target for test projects', () => {
    jest
      .spyOn(fs, 'readFileSync')
      .mockReturnValueOnce('<project ref=Microsoft.NET.Test.Sdk></project>');

    const targets = registerProjectTargets('libs/api/my.csproj');
    expect(targets.test).toMatchSnapshot();
  });
});
