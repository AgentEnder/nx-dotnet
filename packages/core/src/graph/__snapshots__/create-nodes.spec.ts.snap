// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`infer-project should generate build, lint, serve targets for projects 1`] = `
{
  "cache": true,
  "configurations": {
    "production": {
      "configuration": "Release",
    },
  },
  "dependsOn": [
    "^build",
  ],
  "executor": "@nx-dotnet/core:build",
  "options": {
    "configuration": "Debug",
    "noDependencies": true,
  },
  "outputs": [
    "{workspaceRoot}/dist/libs/api",
    "{workspaceRoot}/dist/intermediates/libs/api",
    "{projectRoot}/bin",
    "{projectRoot}/obj",
  ],
}
`;

exports[`infer-project should generate build, lint, serve targets for projects 2`] = `
{
  "cache": true,
  "executor": "@nx-dotnet/core:format",
  "inputs": [
    "{projectRoot}/**/*.{cs,fs,vb}",
  ],
}
`;

exports[`infer-project should generate build, lint, serve targets for projects 3`] = `
{
  "configurations": {
    "production": {
      "configuration": "Release",
    },
  },
  "executor": "@nx-dotnet/core:serve",
  "options": {
    "configuration": "Debug",
  },
}
`;

exports[`infer-project should generate test target for test projects 1`] = `
{
  "cache": true,
  "dependsOn": [
    "build",
  ],
  "executor": "@nx-dotnet/core:test",
  "options": {
    "noBuild": true,
    "testProject": undefined,
  },
}
`;
